lxd/apparmor/archive.go

```diff
@@ -0,0 +1,141 @@
+package apparmor
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/lxc/lxd/lxd/sys"
+	"github.com/lxc/lxd/shared"
+)
+
+var archiveProfileTpl = template.Must(template.New("archiveProfile").Parse(`#include <tunables/global>
+profile "{{.name}}" {
+  #include <abstractions/base>
+  #include <abstractions/nameservice>
+
+{{if .execPath -}}
+  {{ .execPath }} mixr,
+{{- end }}
+{{if .unpackerPath -}}
+  {{ .unpackerPath }} mixr,
+{{- end }}
+
+  {{ .outputPath }}/ rw,
+  {{ .outputPath }}/** rw,
+  {{ .varPath }}/backups/** rw,
+
+  signal (receive) set=("term") peer=unconfined,
+
+  # Capabilities
+  capability dac_read_search,
+  capability dac_override,
+  capability chown,
+  capability fsetid,
+
+{{- if .snap }}
+  # Snap-specific libraries
+  /snap/lxd/*/lib/**.so*                  mr,
+{{- end }}
+}
+`))
+
+// ArchiveLoad ensures that the archive's policy is loaded into the kernel.
+func ArchiveLoad(sysOS *sys.OS, outputPath string, execName string, unpackerName string) error {
+	profile := filepath.Join(aaPath, "profiles", ArchiveProfileFilename(outputPath))
+	content, err := ioutil.ReadFile(profile)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	updated, err := archiveProfile(outputPath, execName, unpackerName)
+	if err != nil {
+		return err
+	}
+
+	if string(content) != string(updated) {
+		err = ioutil.WriteFile(profile, []byte(updated), 0600)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = loadProfile(sysOS, ArchiveProfileFilename(outputPath))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ArchiveUnload ensures that the archive's policy namespace is unloaded to free kernel memory.
+// This does not delete the policy from disk or cache.
+func ArchiveUnload(sysOS *sys.OS, outputPath string) error {
+	err := unloadProfile(sysOS, ArchiveProfileName(outputPath), ArchiveProfileFilename(outputPath))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ArchiveDelete removes the profile from cache/disk
+func ArchiveDelete(sysOS *sys.OS, outputPath string) error {
+	return deleteProfile(sysOS, ArchiveProfileName(outputPath), ArchiveProfileFilename(outputPath))
+}
+
+// archiveProfile generates the AppArmor profile template from the given destination path.
+func archiveProfile(outputPath string, execName string, unpackerName string) (string, error) {
```

Comment:
I think I would prefer us to pass full paths for commands allowed to be run here and not perform `exec.LookPath` in this function, this way its up to the caller to resolve the command(s) that need to be run, as its possible they want to run something that isn't in the path.

Reply:
I've moved `exec.LookPath` outside of `archiveProfile`.